Minor UI changes: Filename only, tooltip, and scrolling

JedBurke opened the pull request:
This pull request contains a few minor user-facing changes:

+ CLI switch to only show the filename, not the full path (enabled with -f/--name-only).
+ Show a tooltip containing the full path when hovering over a file.
+ Use `gtk_scrolled_window_new` to allow scrolling when there are many items.

They've been kept separate as to cherry pick the desired features.

JedBurke committed:
Add option to only show the basename

JedBurke committed:
Show a tooltip of the file's name or path when hovering over the file

JedBurke committed:
Add scrolling behavior to accommodate many files

JedBurke committed:
Align -f option in help text

JedBurke committed:
Document the -f/--name-only option

FichteFoll commented on dragon.c:
```diff
@@ -551,6 +558,9 @@ int main (int argc, char **argv) {
     window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
 
+    // Scrolling Window
+    panel = gtk_scrolled_window_new(NULL, NULL);
```
How is this panel's size determined? As large as can be without exceeding monitor dimensions?

JedBurke replied:
Will have to give it a look. The widget plays nicely in i3 with titling mode (takes the full height) but not in floating.

JedBurke replied:
> How is this panel's size determined?

It takes the size of the parent window. So if the default size is set, `gtk_window_set_default_size(GTK_WINDOW(window), 600, 500)` the scroll window will occupy that space.

> As large as can be without exceeding monitor dimensions?

Might be dependent on the user's desktop environment. Testing on i3 with floating mode, the window size is curbed when setting the default size (with `gtk_window_set_default_size`) to a height greater than the monitor.

mwh replied:
It seems that this results in the smallest possible non-resizable square window (but scrollable). It needs better logic for determining a size, probably only scrolling if it can't fit the contents. I don't know how to make that work.

JedBurke replied:
Yesterday I had a look to set the default size according to the user's screen dimensions without hard coding it. [gtk_window_set_geometry_hints](https://docs.gtk.org/gtk3/method.Window.set_geometry_hints.html) with a comfortable hard coded minimum *could* prevent the window from being too small unless overridden by [gtk_window_set_default_size](https://docs.gtk.org/gtk3/method.Window.set_default_size.html) which was looking to be the result of [gdk_monitor_get_workarea](https://docs.gtk.org/gdk3/method.Monitor.get_workarea.html). 

> probably only scrolling if it can't fit the contents

That's a better idea. The scrollbars already show when necessary. Again, without setting the default window size,  [gtk_scrolled_window_set_min_content_width](https://docs.gtk.org/gtk3/method.ScrolledWindow.set_min_content_width.html) and [gtk_scrolled_window_set_min_content_height](https://docs.gtk.org/gtk3/method.ScrolledWindow.set_min_content_height.html) will set the top-level window's size.

With regards to a dynamic size. Not sure if it's possible, but how about summing the height of each button and using it for the window if it's lower than the working area? The same would go for the width, select the widest button and use its width if it's less than the working area width.

JedBurke replied:
> Not sure if it's possible, but how about summing the height of each button and using it for the window if it's lower than the working area?

It's simpler than that, after calling ` gtk_widget_show_all(window);
`, the allocated width and height of `vbox` can be used. Its height and width can exceed the workarea, so they need to be curbed before calling the respective scrolled_window functions.

Additionally, `gtk_window_set_geometry_hints` can be used to ensure a comfortable minimum size if vbox is too small. In the example below, the left has no minimum size while right has a minimum of 640x480.
![comparison](https://github.com/user-attachments/assets/d498b640-f333-4f4d-94cf-98343fdee1ea)

FichteFoll replied:
I assume this was just an example, but please do not make the window larger than it needs to be. I frequently use dragon with single items and would prefer those not to use 500px in height when 80 would have sufficed.

JedBurke replied:
Yes, this was an example of how to control the size. I will take it under advisement and not enable it.

FichteFoll commented on dragon.c:
```diff
@@ -160,6 +160,9 @@ GtkButton *add_button(char *label, struct draggable_thing *dragdata, int type) {
     }
 
+    // Show a tooltip with the filename. Should perhaps show the full path.
+    gtk_widget_set_tooltip_text(button, label);
```
This is a good idea. I'd even argue that with the tooltip the "name only" option could become the default behavior, perhaps even the only one, but the option is cheap to implement anyway.

JedBurke replied:
If you feel like it can become the default, that's good. The option was included as not to cause disruptions for longtime users. So it can be removed if unnecessary.

mwh replied:
I think the ideal default is probably something like "match name given on command line, if any", so that e.g. `dragon */index.html` preserves the relative path to distinguish items with the same basename, but wouldn't convert everything to absolute paths the way it does now; the architecture makes that tricky currently, I think. There could then be options for basename-only (this commit) and absolute paths-only (current default). This change is a good step for common cases now and it would be a reasonable default for as long as we don't have that middle point. The tooltip makes it possible to distinguish them in the cases where it matters.

Perhaps `-f` like this, `-F`/`--full-path` for the existing default, and switch by default to the `-f` behaviour for the moment?

mwh replied:
I have made it do something like that, at least when the path is within the current directory. It could probably be smarter, but it's a start. Thanks for the PR! These were useful. The tooltip is a great idea.

mwh replied:
I was just thinking that `dragon ../x` would do better to display "../x" instead of an absolute path. Resolving duplicates could be interesting too. I don't expect either of those is actually worthwhile - fiddling around with the paths makes a pretty simple tool a lot more complex.

JedBurke committed:
Set the default size for the top-level window

mwh commented:
Other than the scroll window that doesn't seem usable without something to impose a window size, these all look good to me. I've cherry-picked the other commits, fixuped the space one, and switched the default button to "relative path to ." with this `-f` option and a new `-F` to switch between basename-only (this PR's `-f`) and absolute-only (previous default). That is all merged now.

Scrolling is probably useful for large lists/small windows (especially in a tiling environment), but we need to have the program work "as normal" in standard cases as well and that doesn't happen yet. If there's a clear way to make it auto-size as now if it can fit, that will be good.

JedBurke committed:
Use the file path to get the pixel buffer

When adding a file button and enabling the filename only option, the
`filename` variable becomes unsuitable for obtaining the pixel buffer
from the file.

JedBurke committed:
Add `tooltip` parameter to the `add_button` function

The tooltip needs to display the file's path, but will be unavailable if
the label is passed as the basename only. If the tooltip value is null,
default to the label input.

JedBurke committed:
Set the main window's minimum size to the inner container

JedBurke committed:
Merge branch 'master' of https://github.com/mwh/dragon

JedBurke commented:
Commits 5b48527 and 9c80b7e explain themselves. [31a53a2](https://github.com/mwh/dragon/pull/73/commits/31a53a26d77ea026cf3ca40f62715a947cdd18a9) sets the window geometry based on vbox's size (which is obtained from its children) up to the working area. Maybe there is an overlooked function, but I wasn't able to set the window including its decorations to the working area. To avoid it extending past the workarea, it will be shaved by 5%.

mwh commented:
This all seems to work well to me. I've merged all the new changes into the main branch now.

JedBurke commented:
Thank you.
